src/plugins/relay/api/relay-api-protocol.c

```diff
@@ -663,6 +663,126 @@ RELAY_API_PROTOCOL_CALLBACK(buffers)
     return RELAY_API_PROTOCOL_RC_OK;
 }
 
+/*
+ * Callback for resource "completion".
+ *
+ * Routes:
+ *   POST /api/completion
+ */
+
+RELAY_API_PROTOCOL_CALLBACK(completion)
+{
+    cJSON *json_response, *json_body;
+    cJSON *json_buffer_id, *json_buffer_name;
+    cJSON *json_data, *json_position;
+    const char *ptr_buffer_name, *ptr_data;
+    int position;
+    char str_id[64];
+    struct t_gui_completion *ptr_completion;
+    struct t_gui_buffer *ptr_buffer;
+
+    json_body = cJSON_Parse(client->http_req->body);
+    if (!json_body)
+        return RELAY_API_PROTOCOL_RC_BAD_REQUEST;
+
+    /* Get buffer either by ID or by name */
+    ptr_buffer = NULL;
+    json_buffer_id = cJSON_GetObjectItem(json_body, "buffer_id");
+    if (json_buffer_id)
+    {
+        if (cJSON_IsNumber(json_buffer_id))
+        {
+            snprintf(str_id, sizeof(str_id),
+                     "%lld", (long long)cJSON_GetNumberValue(json_buffer_id));
+            ptr_buffer = weechat_buffer_search("==id", str_id);
+            if (!ptr_buffer)
+            {
+                relay_api_msg_send_error_json(
+                    client,
+                    RELAY_HTTP_400_BAD_REQUEST, NULL,
```

Comment:
400 seems like the correct response, as the route is available (i.e. not 404), but the buffer ID is invalid. this is inconsistent with the handling of the `/api/input` handler (line 849+) though => change input route as well?

Reply:
Done.